File: java/src/tmp/TemplateMT.java
import java.io.File;
import java.io.PrintWriter;
import java.util.Collections;
import java.util.Locale;
import java.util.Scanner;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;

public class TemplateMT {

	public static void main(String[] args) throws Exception {
		long start = System.currentTimeMillis();
		String name = "C-large-practice";
		String path = "";

		Locale.setDefault(Locale.US);

		Scanner sc = new Scanner(new File(path + name + ".in"));
		PrintWriter pw = new PrintWriter(path + name + ".out");
//		Scanner sc = new Scanner(System.in);
//		PrintWriter pw = new PrintWriter(System.out);

		ExecutorService executor = Executors.newFixedThreadPool(16);

		int testCases = sc.nextInt();
		Future<int[]>[] futures = new Future[testCases];
		Set<Integer> set = Collections.newSetFromMap(new ConcurrentHashMap<>());
		for (int testCase = 1; testCase <= testCases; testCase++) {
			int n = sc.nextInt();
			int[] x = new int[n];
			int[] y = new int[n];
			for (int i = 0; i < n; i++) {
				x[i] = sc.nextInt();
				y[i] = sc.nextInt();
			}
			int cur = testCase;

			futures[testCase - 1] = executor.submit(() -> {
				set.add(cur);
				System.out.println(set);
				int[] res = new int[n];
				for (int i = 0; i < n; i++) {
					int cnt = 0;
					for (int j = 0; j < n; j++) {
						if (i == j) continue;
						int x1 = x[j] - x[i];
						int y1 = y[j] - y[i];
						int cnt1 = 0;
						int cnt2 = 0;
						for (int k = 0; k < n; k++) {
							int x2 = x[k] - x[i];
							int y2 = y[k] - y[i];
							long cross = (long) x1 * y2 - (long) x2 * y1;
							if (cross <= 0) ++cnt1;
							if (cross >= 0) ++cnt2;
						}
						cnt = Math.max(cnt, Math.max(cnt1, cnt2));
					}
					res[i] = n == 1 ? 0 : n - cnt;
				}
				set.remove(cur);
				System.out.println(set);
				return res;
			});
		}

		for (int testCase = 1; testCase <= testCases; testCase++) {
			int[] res = futures[testCase - 1].get();
			pw.println("Case #" + testCase + ":");
			for (int v : res) {
				pw.println(v);
			}
			pw.flush();
		}

		pw.close();
		sc.close();
		executor.shutdown();
		System.err.println(System.currentTimeMillis() - start);
	}
}

File: cpp/templateMT.cpp
#pragma comment(linker, "/STACK:268435456")

#include <algorithm>
#include <iostream>
#include <future>

using namespace std;

vector<int> solve(vector<int> x, vector<int> y) {
	int n = x.size();
	vector<int> res(n);
	for (int i = 0; i < n; i++) {
		int cnt = 0;
		for (int j = 0; j < n; j++) {
			if (i == j) continue;
			int x1 = x[j] - x[i];
			int y1 = y[j] - y[i];
			int cnt1 = 0;
			int cnt2 = 0;
			for (int k = 0; k < n; k++) {
				int x2 = x[k] - x[i];
				int y2 = y[k] - y[i];
				long long cross = (long long) x1 * y2 - (long long) x2 * y1;
				if (cross <= 0) ++cnt1;
				if (cross >= 0) ++cnt2;
			}
			cnt = max(cnt, max(cnt1, cnt2));
		}
		res[i] = n == 1 ? 0 : n - cnt;
	}
	return res;
}

int main() {
	auto start = chrono::system_clock::now();

	string name = "C-large-practice";
	string path = "";

	freopen((path + name + ".in").c_str(), "r", stdin);
	freopen((path + name + ".out").c_str(), "w", stdout);

	int test_cases;
	cin >> test_cases;

	future<vector<int>> futures[test_cases];

	for (int test_case = 1; test_case <= test_cases; test_case++) {
		int n;
		cin >> n;
		vector<int> x(n);
		vector<int> y(n);
		for (int i = 0; i < n; i++) {
			cin >> x[i];
			cin >> y[i];
		}

		futures[test_case - 1] = async(launch::async, solve, x, y);
	}

	for (int test_case = 1; test_case <= test_cases; test_case++) {
		vector<int> res = futures[test_case - 1].get();
		cout << "Case #" << test_case << ": " << endl;
		for (int v : res) {
			cout << v << endl;
		}
		cout.flush();
	}

	fclose(stdout);
	fclose(stdin);

	cerr << chrono::duration_cast<chrono::milliseconds>(chrono::system_clock::now() - start).count() << endl;
	return 0;
}
